diff -urN qtbase-opensource-src-5.0.0-beta2.old/mkspecs/features/dbuscommon.pri qtbase-opensource-src-5.0.0-beta2/mkspecs/features/dbuscommon.pri
--- qtbase-opensource-src-5.0.0-beta2.old/mkspecs/features/dbuscommon.pri	2012-11-10 21:03:34.000000000 +0200
+++ qtbase-opensource-src-5.0.0-beta2/mkspecs/features/dbuscommon.pri	2012-11-22 17:11:59.490849805 +0200
@@ -1,5 +1,5 @@
 load(moc)
-qtPrepareTool(QMAKE_QDBUSXML2CPP, qdbusxml2cpp)
+qtPrepareTool(QMAKE_QDBUSXML2CPP, qdbusxml2cpp-qt5)
 
 defineReplace(qdbusOutputBasename) {
     return($$lower($$section($$list($$basename(1)),.,-2,-2)))
diff -urN qtbase-opensource-src-5.0.0-beta2.old/mkspecs/features/resources.prf qtbase-opensource-src-5.0.0-beta2/mkspecs/features/resources.prf
--- qtbase-opensource-src-5.0.0-beta2.old/mkspecs/features/resources.prf	2012-11-10 21:03:34.000000000 +0200
+++ qtbase-opensource-src-5.0.0-beta2/mkspecs/features/resources.prf	2012-11-22 16:55:20.058859693 +0200
@@ -1,4 +1,4 @@
-qtPrepareTool(QMAKE_RCC, rcc)
+qtPrepareTool(QMAKE_RCC, rcc-qt5)
 
 isEmpty(RCC_DIR):RCC_DIR = .
 isEmpty(QMAKE_MOD_RCC):QMAKE_MOD_RCC = qrc

--- a/src/tools/rcc/rcc.pro	2012-12-03 19:18:32.000000000 +0000
+++ b/src/tools/rcc/rcc.pro	2012-12-07 09:52:11.104108391 +0000
@@ -2,6 +2,7 @@
 
 DEFINES += QT_RCC QT_NO_CAST_FROM_ASCII
 
+TARGET=rcc-qt5
 include(rcc.pri)
 HEADERS += ../../corelib/kernel/qcorecmdlineargs_p.h
 SOURCES += main.cpp
--- a/src/tools/qdbusxml2cpp/qdbusxml2cpp.pro	2012-12-03 19:18:32.000000000 +0000
+++ b/src/tools/qdbusxml2cpp/qdbusxml2cpp.pro	2012-12-07 09:53:30.600107750 +0000
@@ -1,6 +1,7 @@
 option(host_build)
 
 DEFINES += QT_NO_CAST_FROM_ASCII
+TARGET = qdbusxml2cpp-qt5
 
 INCLUDEPATH += $$QT_BUILD_TREE/include \
                    $$QT_BUILD_TREE/include/QtDBus \
--- a/src/tools/qdbuscpp2xml/qdbuscpp2xml.pro	2012-12-03 19:18:32.000000000 +0000
+++ b/src/tools/qdbuscpp2xml/qdbuscpp2xml.pro	2012-12-07 09:54:15.412107094 +0000
@@ -3,6 +3,7 @@
 include(../moc/moc.pri)
 
 DEFINES += QT_NO_CAST_FROM_ASCII
+TARGET = qdbuscpp2xml-qt5
 
 INCLUDEPATH += $$QT_BUILD_TREE/include \
                    $$QT_BUILD_TREE/include/QtDBus \
--- a/src/tools/uic/uic.pro	2012-12-03 19:18:32.000000000 +0000
+++ b/src/tools/uic/uic.pro	2012-12-07 09:55:27.900112667 +0000
@@ -1,6 +1,7 @@
 option(host_build)
 
 DEFINES += QT_UIC QT_NO_CAST_FROM_ASCII
+TARGET = uic-qt5
 
 include(uic.pri)
 include(cpp/cpp.pri)
--- a/src/tools/moc/moc.pro	2012-12-03 19:18:32.000000000 +0000
+++ b/src/tools/moc/moc.pro	2012-12-07 09:56:01.092101976 +0000
@@ -1,6 +1,7 @@
 option(host_build)
 
 DEFINES += QT_MOC QT_NO_CAST_FROM_ASCII QT_NO_CAST_FROM_BYTEARRAY QT_NO_COMPRESS
+TARGET = moc-qt5
 
 INCLUDEPATH += $$QT_BUILD_TREE/src/corelib/global
 
