diff --git a/mkspecs/features/qt_headersclean.prf b/mkspecs/features/qt_headersclean.prf
index ab0a722..1e420b0 100644
--- a/mkspecs/features/qt_headersclean.prf
+++ b/mkspecs/features/qt_headersclean.prf
@@ -18,7 +18,7 @@
 # Enable pedantic mode, but accept variadic macros and 'long long' usage.
 *-g++*: QMAKE_CXXFLAGS += -Wno-long-long -Wno-variadic-macros -pedantic-errors
 
-contains(QT_ARCH,arm)|contains(QT_ARCH,mips) {
+contains(QT_ARCH,arm)|contains(QT_ARCH,mips)|contains(QT_ARCH,sparc) {
     # There are outstanding alignment issues in some container classes.
     *-g++*:QMAKE_CXXFLAGS -= -Wcast-align
 }
