Description: enable sparc processor detection
Author: Lisandro Damián Nicanor Pérez Meyer <lisandro@debian.org>
Forwarded: https://codereview.qt-project.org/#change,74515

diff --git a/src/corelib/global/qprocessordetection.h b/src/corelib/global/qprocessordetection.h
index 623e30a..684b7b5 100644
--- a/src/corelib/global/qprocessordetection.h
+++ b/src/corelib/global/qprocessordetection.h
@@ -292,12 +292,12 @@
     SPARC is big-endian only prior to V9, while V9 is bi-endian with big-endian
     as the default byte order. Assume all SPARC systems are big-endian.
 */
-// #elif defined(__sparc__)
-// #  define Q_PROCESSOR_SPARC
-// #  if defined(__sparc_v9__)
-// #    define Q_PROCESSOR_SPARC_V9
-// #  endif
-// #  define Q_BYTE_ORDER Q_BIG_ENDIAN
+#elif defined(__sparc__)
+#  define Q_PROCESSOR_SPARC
+#  if defined(__sparc_v9__)
+#    define Q_PROCESSOR_SPARC_V9
+#  endif
+#  define Q_BYTE_ORDER Q_BIG_ENDIAN
 
 #endif
 
