Description: enable s390[x] processor detection
Author: Lisandro Damián Nicanor Pérez Meyer <lisandro@debian.org>
Forwarded: https://codereview.qt-project.org/#change,74514

diff --git a/src/corelib/global/qprocessordetection.h b/src/corelib/global/qprocessordetection.h
index 623e30a..dde8db0 100644
--- a/src/corelib/global/qprocessordetection.h
+++ b/src/corelib/global/qprocessordetection.h
@@ -267,12 +267,12 @@
 
     S390 is big-endian.
 */
-// #elif defined(__s390__)
-// #  define Q_PROCESSOR_S390
-// #  if defined(__s390x__)
-// #    define Q_PROCESSOR_S390_X
-// #  endif
-// #  define Q_BYTE_ORDER Q_BIG_ENDIAN
+#elif defined(__s390__)
+#  define Q_PROCESSOR_S390
+#  if defined(__s390x__)
+#    define Q_PROCESSOR_S390_X
+#  endif
+#  define Q_BYTE_ORDER Q_BIG_ENDIAN
 
 /*
     SuperH family, optional revision: SH-4A
