Description: QNativeSocketEngine: fix undefined variable on FreeBSD
Author: Dmitry Shachnev <mitya57@debian.org>
Forwarded: https://codereview.qt-project.org/144838
Last-Update: 2015-12-30

--- a/src/network/socket/qnativesocketengine_unix.cpp
+++ b/src/network/socket/qnativesocketengine_unix.cpp
@@ -994,7 +994,7 @@
 #  elif defined(IP_SENDSRCADDR)
             struct in_addr *data = reinterpret_cast<in_addr *>(CMSG_DATA(cmsgptr));
             cmsgptr->cmsg_type = IP_SENDSRCADDR;
-            addr->s_addr = htonl(header.senderAddress.toIPv4Address());
+            data->s_addr = htonl(header.senderAddress.toIPv4Address());
 #  endif
             cmsgptr->cmsg_level = IPPROTO_IP;
             msg.msg_controllen += CMSG_SPACE(sizeof(*data));
